modules/avro-exporter/src/main/scala/io/janstenpickle/trace4cats/avro/AvroSpanExporter.scala

```diff
@@ -69,40 +75,39 @@ object AvroSpanExporter {
     ): F[Unit] =
       Stream
         .repeatEval {
-          (for {
+          for {
             batch <- queue.dequeue1
-            _ <- semaphore.acquire
-            _ <- batch.spans.traverse { span =>
-              for {
-                ba <- encode[F](schema)(span)
-                _ <- socket.write(Packet(address, Chunk.bytes(ba)))
-              } yield ()
+            _ <- semaphore.withPermit {
+              batch.spans.traverse { span =>
+                for {
+                  ba <- encode[F](schema)(span)
+                  _ <- socket.write(Packet(address, Chunk.bytes(ba)))
+                } yield ()
+              }
             }
-          } yield ()).guarantee(semaphore.release)
+          } yield ()
         }
         .compile
         .drain
 
     for {
       avroSchema <- Resource.liftF(AvroInstances.completedSpanSchema[F])
       address <- Resource.liftF(Sync[F].delay(new InetSocketAddress(host, port)))
-      queue <- Resource.liftF(InspectableQueue.bounded[F, Batch[G]](1))
-      semaphore <- Resource.liftF(Semaphore[F](Long.MaxValue))
+      queue <- Resource.liftF(InspectableQueue.bounded[F, Batch[G]](queueCapacity))
+      semaphore <- Resource.liftF(Semaphore[F](maxPermits))
       socketGroup <- UDPSocketGroup(blocker)
       socket <- socketGroup.open()
-      _ <- Resource.make(
+      writer = Resource.make(
         Stream
           .retry(write(avroSchema, address, semaphore, queue, socket), 5.seconds, _ + 1.second, Int.MaxValue)
           .compile
           .drain
           .start
       )(fiber =>
-        Applicative[F].unit.whileM_(for {
-          queueNonEmpty <- queue.getSize.map(_ != 0)
-          semaphoreSet <- semaphore.count.map(_ == 0)
-          _ <- Timer[F].sleep(50.millis)
-        } yield queueNonEmpty || semaphoreSet) >> fiber.cancel
+        Timer[F].sleep(50.millis).whileM_(semaphore.available.map(_ < maxPermits)) >>
```

Comment:
This is probably not the best strategy for releasing multiple fibers, since every single fiber waits until no work is being done at all.